scipy/signal/_bsplines.py

```diff
@@ -470,60 +472,109 @@ def _cubic_smooth_coeff(signal, lamb):
     rho, omega = _coeff_smooth(lamb)
     cs = 1 - 2 * rho * cos(omega) + rho * rho
     K = len(signal)
-    yp = zeros((K,), signal.dtype.char)
     k = arange(K)
-    yp[0] = (_hc(0, cs, rho, omega) * signal[0] +
-             add.reduce(_hc(k + 1, cs, rho, omega) * signal))
-
-    yp[1] = (_hc(0, cs, rho, omega) * signal[0] +
-             _hc(1, cs, rho, omega) * signal[1] +
-             add.reduce(_hc(k + 2, cs, rho, omega) * signal))
-
-    for n in range(2, K):
-        yp[n] = (cs * signal[n] + 2 * rho * cos(omega) * yp[n - 1] -
-                 rho * rho * yp[n - 2])
-
-    y = zeros((K,), signal.dtype.char)
-
-    y[K - 1] = add.reduce((_hs(k, cs, rho, omega) +
-                           _hs(k + 1, cs, rho, omega)) * signal[::-1])
-    y[K - 2] = add.reduce((_hs(k - 1, cs, rho, omega) +
-                           _hs(k + 2, cs, rho, omega)) * signal[::-1])
-
-    for n in range(K - 3, -1, -1):
-        y[n] = (cs * yp[n] + 2 * rho * cos(omega) * y[n + 1] -
-                rho * rho * y[n + 2])
 
+    zi_2 = (_hc(0, cs, rho, omega) * signal[0] +
+            add.reduce(_hc(k + 1, cs, rho, omega) * signal))
+    zi_1 = (_hc(0, cs, rho, omega) * signal[0] +
+            _hc(1, cs, rho, omega) * signal[1] +
+            add.reduce(_hc(k + 2, cs, rho, omega) * signal))
+
+    # Forward filter:
+    # for n in range(2, K):
+    #     yp[n] = (cs * signal[n] + 2 * rho * cos(omega) * yp[n - 1] -
+    #              rho * rho * yp[n - 2])
+    zi = lfiltic(cs, r_[1, -2 * rho * cos(omega), rho * rho], r_[zi_1, zi_2])
+    zi = atleast_2d(zi)
+
+    sos = r_[cs, 0, 0, 1, -2 * rho * cos(omega), rho * rho]
+    sos = atleast_2d(sos)
```

Comment:
Minor, non-blocking: maybe `np.r_[....].reshape(1, -1)` instead of atleast_2d would be a bit more explicit/readable.